### PR TITLE
Assignment 3 - word search

### PR DESCRIPTION
**Pull request by ElizabethSvit**

### CODE REVIEW

File: ass3.cpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
+#include <iostream>
+#include <cstring>
+#include <set>
+#include <vector>
+
+using namespace std;
+
+#define ROW_SIZE 2
```
Comment (ssancho2) [inline ass3.cpp]: Having defines for this is too limiting, you should be able to exercise the code with grids of different sizes.

File: ass3.cpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+#define COL_SIZE 3
+
+
+class Dictionary {
+    set<string> dict_words;
+    set<string> dict_prefixes;
+public:
+    Dictionary() {};
+    
+    void fill_dict(vector<string> words) {
```
Comment (ssancho2) [inline ass3.cpp]: The vector is being copied here, no need for that, better this to be a const &

File: ass3.cpp [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        for (int i = 0; i < words.size(); ++i) {
```
Comment (ssancho2) [inline ass3.cpp]: You can do:
`for (const string &word : words) { ... }`

File: ass3.cpp [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+            dict_words.insert(words[i]);
+            
+            string prefix = "";
+            for (int j = 0; j < words[i].length(); ++j) {
```
Comment (ssancho2) [inline ass3.cpp]: And then here you can do:
`for (char c : word) { ... }`

File: ass3.cpp [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+                prefix += words[i][j];
+                dict_prefixes.insert(prefix);
+            }
+        }
+    }
+
+
+    bool isWord(string s) {
```
Comment (ssancho2) [inline ass3.cpp]: Better to avoid a copy, have the parameter be const &

File: ass3.cpp [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        return dict_words.find(s) != dict_words.end();
+    }
+    
+    bool isPrefix(string s) {
```
Comment (ssancho2) [inline ass3.cpp]: Same here.

File: ass3.cpp [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        return dict_prefixes.find(s) != dict_prefixes.end();
+    }
+};
+
+void find_pref_word(char table[ROW_SIZE][COL_SIZE], vector<vector<bool> > visited, int i, int j, string &str, Dictionary dict, set<string> &result_words) {
```
Comment (ssancho2) [inline ass3.cpp]: I think the table should be `vector<vector<char>>` or `vector<string>`.
If you want, you can use `char *` or `char **` and pass the sizes as parameters.

File: ass3.cpp [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    visited[i][j] = true;
```
Comment (ssancho2) [inline ass3.cpp]: You can move this to around line 49 and avoid `visited[i][j] = false;` on line 46 :)

File: ass3.cpp [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+                find_pref_word(table, visited, row, col, str, dict, result_words);
+            }
+        }
+    }
+
+    str.erase(str.length() - 1);
+    visited[i][j] = false;
+}
+
+void find_all_words(char table[ROW_SIZE][COL_SIZE], Dictionary dict, set<string> &result_words) {
```
Comment (ssancho2) [inline ass3.cpp]: Make dict a const &

File: ass3.cpp [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    vector<vector<bool> > visited;
```
Comment (ssancho2) [inline ass3.cpp]: You can initialize directly with:
`vector<vector<bool> > visited(ROW_SIZE, vector<bool>(COL_SIZE, false));`

File: ass3.cpp [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+int main(int argc, const char * argv[]) {
+    Dictionary dict;
+    dict.fill_dict({"CAR", "CARD", "CART", "CAT"});
+
+   char table[ROW_SIZE][COL_SIZE] = {{'A','A','R'}, {'T','C','D'}};
+
+    set<string> result_words;
+    find_all_words(table, dict, result_words);
+
+    for (auto it = result_words.begin(); it != result_words.end(); ++it) {
```
Comment (ssancho2) [inline ass3.cpp]: Can do:
`for (const string &word : result_words) { ... } `

File: ass3.cpp (hunk 7 above)
Comment (ssancho2) [inline ass3.cpp]: Also, I think visited should be a reference.

Comment (ssancho2) [inline ass3.cpp]: And `i` and `j` are not very clear names... I think either `row` & `col` or `x` & `y` are better.

**Review by ssancho2 — COMMENTED** [review]
Nice concise code!

**Comment by ElizabethSvit** [comment]
Thank you for your comments! I've updated the code and commited it here. 
However, I suppose I cannot request review again unless every reviewer writes comments.

File: ass3.cpp [hunk 12]
```diff
@@ -5,75 +5,70 @@
...
     }
     
+    visited[y][x] = true;
+    
     if (dict.isWord(str))
         result_words.insert(str);
     
-    for (int row = i - 1; row <= i + 1 && row < ROW_SIZE; ++row) {
-        for (int col = j - 1; col <= j + 1 && col < COL_SIZE; ++col) {
+    for (int row = y - 1; row <= y + 1 && row < row_size; ++row) {
```
Comment (ssancho2) [inline ass3.cpp]: NIT: try to stick to a convention, either row & col or x & y.

File: ass3.cpp [hunk 13]
```diff
@@ -5,75 +5,70 @@
...
+    visited[y][x] = false;
 }
 
-void find_all_words(char table[ROW_SIZE][COL_SIZE], Dictionary dict, set<string> &result_words) {
-    vector<vector<bool> > visited;
-    vector<bool> row_visited(COL_SIZE, false);
-    for (int i = 0; i < ROW_SIZE; ++i) {
-        visited.push_back(row_visited);
-    }
+void find_all_words(vector<vector<char>> table, Dictionary const &dict, set<string> &result_words) {
```
Comment (ssancho2) [inline ass3.cpp]: Not a big deal here, but table will be copied, better to mark it as `const &`

File: ass3.cpp [hunk 14]
```diff
@@ -5,75 +5,70 @@
...
-    bool isPrefix(string s) {
+    bool isPrefix(string const &s) {
         return dict_prefixes.find(s) != dict_prefixes.end();
     }
 };
 
-void find_pref_word(char table[ROW_SIZE][COL_SIZE], vector<vector<bool> > visited, int i, int j, string &str, Dictionary dict, set<string> &result_words) {
-    visited[i][j] = true;
-    str += table[i][j];
+void find_pref_word(vector<vector<char>> &table, vector<vector<bool> > &visited, int y, int x, string &str, Dictionary dict, set<string> &result_words) {
```
Comment (ssancho2) [inline ass3.cpp]: Maybe add `const` to table?

**Review by ssancho2 — APPROVED** [review]
Looks good, minor comments. Fell free to merge as it is and move to the next one.

**Comment by ssancho2** [comment]
And , no problem, happy that you find it helpful.
No need to ask for another review, I get notified that you posted new commits and can review again.

**Pull request merged**